src/MultiFormats/MultiAddress.cs

```diff
@@ -0,0 +1,565 @@
+        /// </remarks>
+        private void Read(CodedInputStream stream)
+        {
+            Protocols.Clear();
+            do
+            {
+                var code = (uint) stream.ReadInt64();
+                if (!NetworkProtocol.Codes.TryGetValue(code, out var protocolType))
+                {
+                    throw new InvalidDataException($"The IPFS network protocol code '{code}' is unknown.");
+                }
+                
+                var p = (NetworkProtocol) Activator.CreateInstance(protocolType);
+                p.ReadValue(stream);
+                Protocols.Add(p);
+            } while (!stream.IsAtEnd);
+        }
+
+        /// <summary>
+        ///   Reads the string representation from the specified <see cref="TextReader"/>.
+        /// </summary>
+        /// <param name="stream">
+        ///   The <see cref="TextReader"/> to read from
+        /// </param>
+        /// <remarks>
+        ///   The string representation is a sequence of <see cref="NetworkProtocol">network protocols</see>.
+        /// </remarks>
+        private void Read(TextReader stream)
+        {
+            if (stream.Read() != '/')
+            {
+                throw new FormatException("An IPFS multiaddr must start with '/'.");
+            }
+
+            var name = new StringBuilder();
+            Protocols.Clear();
+            int c;
+            while (true)
+            {
+                name.Clear();
+                while (-1 != (c = stream.Read()) && c != '/')
+                {
+                    name.Append((char) c);
+                }
+
+                if (name.Length == 0)
+                {
+                    break;
+                }
+                
+                if (!NetworkProtocol.Names.TryGetValue(name.ToString(), out var protocolType))
+                {
+                    throw new FormatException(string.Format("The IPFS network protocol '{0}' is unknown.",
+                        name));
+                }
+                
+                var p = (NetworkProtocol) Activator.CreateInstance(protocolType);
+                p.ReadValue(stream);
+                Protocols.Add(p);
+            }
+
+            if (Protocols.Count == 0)
+            {
+                throw new FormatException("The IFPS multiaddr has no protocol specified.");
+            }
+        }
+
+        /// <inheritdoc />
+        public override int GetHashCode()
+        {
+            var code = 0;
+
+            foreach (var p in Protocols)
+            {
+                code += p.Code.GetHashCode();
+                code += p.Value?.GetHashCode() ?? 0;
+            }
+
+            return code;
+        }
+
+        /// <inheritdoc />
+        public override bool Equals(object obj)
+        {
+            var that = obj as MultiAddress;
+            return that != null && Equals(that);
+        }
+
+        /// <inheritdoc />
+        public bool Equals(MultiAddress that)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Rename parameter 'that' to 'other' to match the interface declaration.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)